File: python/tests/shop_driver.py
"""Chapter 6: the acceptance-test driver. Tests speak domain; HTTP lives here."""
from fastapi.testclient import TestClient

from shop import (CatalogPricing, InMemoryInventory, InMemoryPlacedOrders,
                  PlacedOrder, build_app)


class ShopDriver:
    def __init__(self):
        self.prices, self.inventory = {}, InMemoryInventory()
        app = build_app(
            orders=InMemoryPlacedOrders(),
            pricing=CatalogPricing(self.prices),
            inventory=self.inventory,
        )
        self.client = TestClient(app)

    def stocks(self, sku, *, price_cents, bags):
        self.prices[sku] = price_cents
        self.inventory.stock(sku, bags)

    def place_order(self, *, customer_id, items):
        lines = [{"sku": s, "quantity": q} for s, q in items]
        res = self.client.post("/orders", json={
            "customer_id": customer_id, "items": lines,
        })
        assert res.status_code == 201, res.text
        return PlacedOrder(**res.json())

    def try_to_place_order(self, *, customer_id, items):
        res = self._post(customer_id, items)
        return res.status_code, res.json()

    def _post(self, customer_id, items):
        lines = [{"sku": s, "quantity": q} for s, q in items]
        return self.client.post("/orders", json={
            "customer_id": customer_id, "items": lines,
        })

File: test/acceptance/shopDriver.ts
// test/acceptance/shopDriver.ts (Chapter 6, final shape)
import { buildApp } from "../../src/app";
import type { Order, OrderRequest } from "../../src/orders";
import { CatalogPricing } from "../../src/pricing";
import {
  FakePaymentGateway,
  InMemoryInventory,
  InMemoryOrderRepository,
  SequentialIds,
} from "../../src/testing/fakes";

type ShopOrder = Omit<OrderRequest, "paymentToken">;

export class ShopDriver {
  private orders = new InMemoryOrderRepository();
  private prices = new Map<string, number>();
  private inventory = new InMemoryInventory();
  private app = buildApp({
    orders: this.orders,
    pricing: new CatalogPricing(this.prices),
    inventory: this.inventory,
    gateway: new FakePaymentGateway(),
    ids: new SequentialIds(),
  });

  stocks(sku: string, s: { priceCents: number; bags: number }) {
    this.prices.set(sku, s.priceCents);
    this.inventory.stock(sku, s.bags);
  }

  async placeOrder(request: ShopOrder): Promise<Order> {
    const res = await this.post(request);
    if (res.statusCode !== 201) {
      throw new Error(
        `POST /orders returned ${res.statusCode}: ${res.body}`,
      );
    }
    return res.json();
  }

  async tryToPlaceOrder(
    request: ShopOrder,
  ): Promise<{ reason: string }> {
    const res = await this.post(request);
    if (res.statusCode === 201) {
      throw new Error("expected the order to be rejected");
    }
    return res.json();
  }

  async hasRecorded(orderId: string): Promise<boolean> {
    return (await this.orders.findById(orderId)) !== null;
  }

  private post(request: ShopOrder) {
    return this.app.inject({
      method: "POST",
      url: "/orders",
      payload: { ...request, paymentToken: "tok_test" },
    });
  }
}
